CI/CD recommendations

jhadobe opened the pull request:
This PR explains how developers can set up a CI/CD pipeline for API Mesh using GitHub.

[staging link](https://adobedocs.github.io/graphql-mesh-gateway/gateway/cicd/)

naomirobichaud commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,237 @@
+
+# CI/CD for API Mesh
+
+API Mesh for Adobe Developer App Builder now provides CI/CD through [Github actions](https://docs.github.com/en/actions). This allows developers to build, test, and deploy meshes faster and more reliably.
```
We should include a note that explains that API Mesh and App Builder now leverage the same tools to enable CI/CD through Github Actions, so the experience should be consistent for both products.

naomirobichaud commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,237 @@
+
+To get an OAuth token to enable CI/CD:
+
+1. Navigate to the [Adobe Developer Console](https://developer.adobe.com/console) and select your desired workspace.
```
I think we should should say select the workspace where your mesh is currently, to avoid confusion

naomirobichaud commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,237 @@
+
+- Variables are for information that is not sensitive, for example, if you are targeting `stage` or `prod`. When viewing your mesh in the Developer Console or by running the `aio api-mesh:get` command, these variables will be exposed.
+
+- Secrets are for sensitive information, such as access tokens. These are never exposed in the Developer Console or when running the `aio api-mesh:get` command.
```
@revanth0212 is this statement correct in current version?

revanth0212 replied:
No @naomirobichaud I believe the wording needs to be adjusted to better match our current offering.

There could be confusion here because we are talking about Github Variables and Github Secrets (not API Mesh) and unfortunately we will be using the same naming convention for our future feature **Secrets Management**.

This statement will be valid once we launch our feature later this year.

Meanwhile, I believe we can adjust the wording to only talk about Github Secrets.

```suggestion
- Secrets are for sensitive information, such as access tokens. These are never exposed in the Github code repo.
```

naomirobichaud commented:
Overall docs look pretty good. I think we need to consider how the API Mesh docs structure scales with topics like CI/CD, Secrets, Logging, and future functionality. 

We also need to make sure that the secrets functionality in this CI/CD topic is accurate today. I'll need @prabhuram93 help to validate that

prabhuram93 commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,237 @@
+
+1. Click **OAuth Server-to-Server** inside the I/O Management API.
+
+1. Click "Generate access token". Copy and save this token for later use.
```
The developers need not generate the token. The GitHub workflow will generate the token based on ClientID, Client secret etc.

prabhuram93 replied:
[This](https://developer.adobe.com/developer-console/docs/guides/authentication/ServerToServerAuthentication/implementation/#generate-access-tokens) is a good article to reference, to generate the access token programmatically. Very useful in the case of BYO CI/CD.

jhadobe commented:
> Overall docs look pretty good. I think we need to consider how the API Mesh docs structure scales with topics like CI/CD, Secrets, Logging, and future functionality.
> 
> We also need to make sure that the secrets functionality in this CI/CD topic is accurate today. I'll need @prabhuram93 help to validate that

The secrets mentioned on this page are GitHub secrets, not controlled by API Mesh.

jhadobe commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,237 @@
+
+API Mesh for Adobe Developer App Builder now provides CI/CD through [Github actions](https://docs.github.com/en/actions). This allows developers to build, test, and deploy meshes faster and more reliably.
+
+## API Mesh authentication
```
I've added some numbering. @prabhuram93  Could you clarify what you mean when comparing it to the App Builder docs?

prabhuram93 replied:
Numbering works. I liked the way how this has been split into Lessons. It made it easy to comprehend. It may not apply to this, so just a recommendation to consider the possibility.

revanth0212 commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,239 @@
+
+When developing locally, you can store your variables and secrets in your [environment variables file](./developer-tools.md#environment-variables). When using GitHub, you will need to use [secrets](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions) and [variables](https://docs.github.com/en/actions/learn-github-actions/variables).
+
+- Variables are for information that is not sensitive, for example, if you are targeting `stage` or `prod`. When viewing your mesh in the Developer Console or by running the `aio api-mesh:get` command, these variables will be exposed.
```
Same here, I believe we are talking about Github Varilabels which don't have any effect on API Mesh variables unless used in the CICD manually. If that is the case, they will be visible in the mesh. Same is applicable to Secrets as well. This is because API Mesh does not have the concept of Variables and Secrets today. Its all Variables. Once we launch Secrets Management we can call out the distinction in our docs. Meanwhile it would be less confusing if we can limit this section to just Github.

```suggestion
- Variables are for information that is not sensitive, for example, if you are targeting `stage` or `prod` and a variable could carry a different value based on the environment.
```

revanth0212 reviewed (changes requested):
Good question @naomirobichaud 

I have added my views on the comment. My suggestions are just a draft, @jhadobe please update my suggestion to better suit the doc language.

jhadobe commented:
> Good question @naomirobichaud
> 
> I have added my views on the comment. My suggestions are just a draft, @jhadobe please update my suggestion to better suit the doc language.

thanks @revanth0212! i've added your suggestions. i believe i was intending to communicate that because the secrets are in GH, it wouldn't be possible to expose them on our side, but i think it makes more sense to keep the discussion on this page focused on the GH side.

prabhuram93 commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,239 @@
+          IMSORGID: ${{ secrets.IMSORGID_PROD }}
+          SCOPES: AdobeID, openid, read_organizations, additional_info.projectedProductContext, additional_info.roles, adobeio_api, read_client_secret, manage_client_secrets
+      - name: Select org
+        run: aio console:org:select 123456
```
This needs to be updated to retrieve from the GitHub secrets. Also applicable for project ID and workspace ID
aio console:org:select ${{ secrets.IMSORGID_PROD }}. 

Should this section be swapped with #4? Since configuring the secrets is a prerequisite to this step.

prabhuram93 replied:
Yes, correct. The workflow templates should not have hardcoded values unless the clients would like to customize them.

keharper commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,242 @@
+
+For example, if you named a secret `CLIENTID_PROD`, you can call that secret using the `${{ secrets.CLIENTID_PROD }}` format.
+
+The specific secrets required for this workflow are available in the [Developer Console](https://console.adobe.io/) by navigating to the desired workspace within a project and clicking the **Download all** button in the top-right of the screen. This downloads a `JSON` file.
```
Is `https://console.adobe.io/` the preferred URL? I thought that domain was replaced with `developer.adobe.com`.

keharper commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,242 @@
+
+<InlineAlert variant="info" slots="text"/>
+
+Before you can successfully run this workflow, you must configure your [GitHub secrets](#4-configure-your-secrets-and-variables).
```
Do you need is note? It was done in the previous step. Also, the link should start with 3. Didn't check if the whole string is still correct.

keharper commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,242 @@
+
+This workflow relies on two existing Adobe I/O GitHub Actions published on the GitHub Marketplace:
+
+-  `adobe/aio-cli-setup-action` installs and configures the [CLI](https://github.com/adobe/aio-cli) on the GitHub infrastructure. For API Mesh, this action handles authentication.
```
Add a link to these actions?

keharper commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,242 @@
+
+## Bring your own CI/CD
+
+Although GitHub is the recommended CI/CD platform, it is possible to use other CI/CD platforms. You will need to refer to the CI/CD platform's documentation for secrets management and workflow creation, because the syntax of the workflow and variables will be different for each platform.
```
```suggestion
Although GitHub is the recommended CI/CD platform, you can use other CI/CD platforms. Refer to the alternative CI/CD platform's documentation for secrets management and workflow creation, because the syntax of the workflow and variables will be different for each platform.
```

prabhuram93 commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,240 @@
+        uses: actions/setup-node@v4
+        with:
+          node-version: ${{ matrix.node-version }}
+      - name: npm install
```
@jhadobe I realized that we do not need an npm install. You can remove this step.

prabhuram93 commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,240 @@
+          IMSORGID: ${{ secrets.IMSORGID_PROD }}
+          SCOPES: AdobeID, openid, read_organizations, additional_info.projectedProductContext, additional_info.roles, adobeio_api, read_client_secret, manage_client_secrets
+      - name: Select org
+        run: aio console:org:select ${{ secrets.ORGID_PROD }}
```
We may need to help differentiate ORGID_PROD and IMSORGID_PROD with examples, as both are not same values. ORGID_PROD is the ID and IMSORGID_PROD is the code.

jhadobe replied:
@prabhuram93 two questions here looking at the download file from dev console: 

1. By `code` do you mean the `"code": "Adobe...",` in the `services` object?

2. Which ID in the following example is the appropriate org id for ORGID_PROD, `id` or `ims_org_id`?

```
  "org": {
      "id": "123456",
      "name": "adobe",
      "ims_org_id": "98765432123456789@AdobeOrg"
  },
```

prabhuram93 replied:
The ORGID_PROD should be "id": "123456"

prabhuram93 commented on src/pages/gateway/cicd.md:
```diff
@@ -0,0 +1,240 @@
+          echo API_ENDPOINT="https://xxx.execute-api.us-west-2.amazonaws.com" >> .env
+          echo API_KEY=${{ secrets.API_KEY }} >> .env
+          cat .env
+      - name: api-mesh update
```
Since we are mentioning just the update, we should add a note to manually create a mesh on the workspace before setting up the CI/CD. The Expectation is to have Mesh on the workspace on which CI/CD is being set up. 

cc: @naomirobichaud This is a feature gap. We need to do some work to get around this.

jhadobe replied:
@prabhuram93 this is mentioned in a note at the beginning of the page: https://github.com/AdobeDocs/graphql-mesh-gateway/pull/207/files#diff-fc6e4143b29f528c5cc247fea51e30b73af74eb31d7a71e498166362fc3316a8R27

prabhuram93 replied:
Sounds good.
